scripts/casefold_db.py

```diff
@@ -0,0 +1,293 @@
+        )
+
+        if casefold_address in associations:
+            associations[casefold_address].append(
+                (address, mxid, lookup_hash, sg_assoc)
+            )
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash, sg_assoc)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[Any, str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
+
+    # list of mxids and addresses to send emails to letting them know the mxid
+    # has been deleted
+    mxids: List[Tuple[Any, Any]] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][3],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for address, mxid, _, _ in assoc_tuples[1:]:
+                to_delete.append((address,))
+                mxids.append((mxid, address))
+
+    # iterate through the mxids and send email, let's only send on email per mxid
+    if flag == "no_email" or flag == "dry_run":
+        pass
+    else:
+        for mxid, address in mxids:
+            processed_mxids = []
+
+            if mxid in processed_mxids:
+                continue
+            else:
+                templateFile = sydent.get_branded_template(
+                    "matrix-org",
+                    "migration_template.eml",
+                    ("email", "email.template"),
+                )
+
+                sendEmail(
+                    sydent,
+                    templateFile,
+                    address,
+                    {"mxid": "mxid", "subject_header_value": "MatrixID Update"},
+                )
+                processed_mxids.append(mxid)
+
+    print(
+        f"{len(to_delete)} rows to delete, {len(db_update_args)} rows to update in global_threepid_associations"
+    )
+
+    if flag == "dry_run":
+        pass
+    else:
+        if len(to_delete) > 0:
+            cur.executemany(
+                "DELETE FROM global_threepid_associations WHERE address = ?", to_delete
+            )
+
+        if len(db_update_args) > 0:
+            cur.executemany(
+                "UPDATE global_threepid_associations SET address = ?, lookup_hash = ?, sgAssoc = ? WHERE address = ? AND mxid = ?",
+                db_update_args,
+            )
+
+        db.commit()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Casefold email addresses in database")
+    parser.add_argument(
+        "--no_email", action="store_true", help="run script but do not send emails"
+    )
+    parser.add_argument("--apply", action="store_true", help="run full script")
```

Comment:
I think we'd be better off with a required (positional) argument where you specify the configuration path. Relying on an environment variable (as `get_config_file_path` does) is OK-ish for a service, but for a command-line tool I think it'd be nicer to be explicit.

example usage would then be `casefold_db.py [--dry-run] [--no-email] sydent/sydent.conf`

Don't worry about giving warnings or anything if both dry-run and no-email are specified; dry-run should imply no-email but won't conflict with it.

Then an admin could first do something like:

* `casefold_db.py sydent/sydent.conf --no-email --dry-run`
* *wait for results, think that it seems reasonable*
* `casefold_db.py sydent/sydent.conf --no-email` (i.e. the same command with only `--dry-run` removed)

If you make the two conflict, you're making it more likely that this will happen:

* `casefold_db.py sydent/sydent.conf --no-email --dry-run` (fail: the flags conflict!!)
* *argh!*
* `casefold_db.py sydent/sydent.conf --dry-run`
* *wait for results, have a cup of tea, come back, think that it seems reasonable*
* `casefold_db.py sydent/sydent.conf` (i.e. the same command with only `--dry-run` removed)
* → oops, forgot the `--no-email` flag that we wanted

(and since we require specifying the config file, we won't need `--apply` anymore since it's already explicit enough that this command will do something)